packages/database/src/migrations/20211128090124-MergeFlutrackingResponses-modifies-data.js

```diff
@@ -0,0 +1,189 @@
+  const answeredQuestionIds = mostRecentAnswers.map(a => a.question_id);
+
+  const responseIdsToDelete = [];
+  const answersToCreate = [];
+  for (let i = 1; i < sortedResponses.length; i++) {
+    const response = sortedResponses[i];
+    const answers = answersByResponseId[response.id];
+    responseIdsToDelete.push(response.id);
+
+    answers.forEach(answer => {
+      const { question_id: questionId } = answer;
+      if (!answeredQuestionIds.includes(questionId)) {
+        answeredQuestionIds.push(questionId);
+        // Merge the answer from a response other than the most recent in the most recent response
+        // All other responses (and their answers) can be deleted since they are redundant
+        answersToCreate.push({ ...answer, survey_response_id: mostRecentResponse.id });
+      }
+    });
+  }
+
+  return { responseIdsToDelete, answersToCreate };
+};
+
+const selectAnswersForResponseGroups = async (db, responseGroups) => {
+  const responseIds = responseGroups.map(group => group.map(r => r.id)).flat();
+  const { rows: answers } = await db.runSql(`
+    select * from answer where survey_response_id IN (${arrayToDbString(responseIds)});
+  `);
+  return groupBy(answers, 'survey_response_id');
+};
+
+const processResponseGroups = async (db, responseGroupBatch) => {
+  const a = Date.now();
+
+  const answerByResponseId = await selectAnswersForResponseGroups(db, responseGroupBatch);
+  const mergeData = responseGroupBatch.map(responseGroup =>
+    mergeAnswersInResponseGroup(responseGroup, answerByResponseId),
+  );
+
+  const responseIdsToDelete = mergeData.map(r => r.responseIdsToDelete).flat();
+  const answersToCreate = mergeData.map(d => d.answersToCreate).flat();
+  await deleteResponses(db, responseIdsToDelete);
+  await createAnswers(db, answersToCreate);
+};
+
+const groupResponsesByEntityAndWeek = duplicateResponses => {
+  const groupedResponses = {};
+
+  duplicateResponses.forEach(record => {
+    const {
+      id1,
+      id2,
+      end_time1: endTime1,
+      end_time2: endTime2,
+      entity_id: entityId,
+      iso_week: isoWeek,
+    } = record;
+    const key = `${entityId}___${isoWeek}`;
+
+    if (!(key in groupedResponses)) {
+      groupedResponses[key] = {};
+    }
+    groupedResponses[key][id1] = { id: id1, endTime: endTime1 };
+    groupedResponses[key][id2] = { id: id2, endTime: endTime2 };
+  });
+
+  return Object.values(groupedResponses).map(group => Object.values(group));
+};
+
+const selectDuplicateFlutrackingResponses = async (db, surveyCode) => {
+  const { rows } = await db.runSql(`
+    select
+      sr.id as id1,
+      sr2.id as id2,
+      sr.end_time as end_time1,
+      sr2.end_time as end_time2,
+      sr.entity_id,
+      concat(date_part('year', sr.data_time), 'W', date_part('week', sr.data_time)) as iso_week
+    from survey_response sr
+    join survey_response sr2 on sr2.id > sr.id and sr2.survey_id = sr.survey_id and sr2.entity_id = sr.entity_id
```

Comment:
`>` is such a clever way of only getting the duplicates in one direction, not double matching!